src/Castle.Core/DynamicProxy/ProxyUtil.cs

```diff
@@ -109,17 +150,134 @@ public static bool IsAccessible(MethodBase method, out string message)
+		public static bool IsAccessible(Type type)
+		{
+			return IsAccessibleType(type);
+		}
+
+		internal static bool IsAccessibleType(Type target)
+		{
+			var typeInfo = target.GetTypeInfo();
+
+			var isPublic = typeInfo.IsPublic || typeInfo.IsNestedPublic;
+			if (isPublic)
+			{
+				return true;
+			}
+
+			var isTargetNested = target.IsNested;
+			var isNestedAndInternal = isTargetNested && (typeInfo.IsNestedAssembly || typeInfo.IsNestedFamORAssem);
+			var isInternalNotNested = typeInfo.IsVisible == false && isTargetNested == false;
+			var isInternal = isInternalNotNested || isNestedAndInternal;
+			if (isInternal && AreInternalsVisibleToDynamicProxy(typeInfo.Assembly))
+			{
+				return true;
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		///   Checks whether the specified method is accessible to DynamicProxy.
+		///   Unlike with <see cref="IsAccessible(MethodBase)"/>, the declaring type's accessibility is ignored.
+		/// </summary>
+		/// <param name = "method">The method to check.</param>
+		/// <returns><c>true</c> if the method is accessible to DynamicProxy, <c>false</c> otherwise.</returns>
+		internal static bool IsAccessibleMethod(MethodBase method)
+		{
+			// Accessibility supported by the full framework and CoreCLR
+			if (method.IsPublic || method.IsFamily || method.IsFamilyOrAssembly)
+			{
+				return true;
+			}
+
+			if (method.IsFamilyAndAssembly)
+			{
+				return true;
+			}
+
+			if (AreInternalsVisibleToDynamicProxy(method.DeclaringType.GetTypeInfo().Assembly) && method.IsAssembly)
+			{
+				return true;
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		///   Determines whether the specified method is internal.
+		/// </summary>
+		/// <param name = "method">The method.</param>
+		/// <returns>
+		///   <c>true</c> if the specified method is internal; otherwise, <c>false</c>.
+		/// </returns>
+		internal static bool IsInternal(MethodBase method)
+		{
+			return method.IsAssembly || (method.IsFamilyAndAssembly && !method.IsFamilyOrAssembly);
+		}
+
+		/// <summary>
+		/// Provides instructions that a user could follow to make a type or method in <paramref name="targetAssembly"/>
+		/// visible to DynamicProxyGenAssembly2.</summary>
+		/// <param name="targetAssembly">The assembly containing the type or method.</param>
+		/// <returns>Instructions that a user could follow to make a type or method visible to DynamicProxyGenAssembly2.</returns>
+		internal static string CreateInstructionsToMakeVisible(Assembly targetAssembly)
+		{
+			string strongNamedOrNotIndicator = " not"; // assume not strong-named
+			string assemblyToBeVisibleTo = "\"DynamicProxyGenAssembly2\""; // appropriate for non-strong-named
+
+			if (targetAssembly.IsAssemblySigned())
+			{
+				strongNamedOrNotIndicator = "";
+				assemblyToBeVisibleTo = ReferencesCastleCore(targetAssembly)
+					? "InternalsVisible.ToDynamicProxyGenAssembly2"
+					: '"' + InternalsVisible.ToDynamicProxyGenAssembly2 + '"';
```

Comment:
I'd submit that we do need it. Okay, maybe not "need". The bit after `?` provides a convenience mechanism for assemblies that reference Castle.Core. They can just add 

```c#
[assembly: InternalsVisibleTo("InternalsVisible.ToDynamicProxyGenAssembly2")]
```

instead of providing the full reference to the assembly themselves. The bit after the `:` gives a message like

```
Can not create proxy for type FluentAssertions.Common.NullReflector because it is not accessible. Make it public, or internal and mark your assembly with [assembly: InternalsVisibleTo("DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7")] attribute, because assembly FluentAssertions.Core is strong-named.
```

a